File: flutter_ui/lib/data/repositories/patients/patient_repository.dart
/*
 * Copyright 2026 IVIR Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'package:flutter_ui/shared/logging/logger.dart';
import 'package:flutter_ui/data/model/patients/patient.dart';
import 'package:flutter_ui/data/provider/patients/patient_rest.dart';

Logger _logger = Logger("PatientRepository");

class PatientRepository {
  final PatientRest patientRest;
  final Map<String, Patient> patientMap = {};

  PatientRepository(this.patientRest);

  Patient? getPatient(String id) {
    return patientMap[id];
  }

  Future<Patient> addNewPatient(Patient newPatient) async {
    _logger.logTrace(1, 'Adding new patient to REST API');
    Patient postedPatient = await patientRest.postNewPatient(newPatient);
    putPatient(postedPatient);
    _logger.logDebug(1, "New patient added: ${postedPatient.id}");
    return postedPatient;
  }

  void putPatient(Patient patient) {
    patientMap[patient.id] = patient;
  }

  Future<List<Patient>> fetchAllPatients() async {
    final List<Patient> restList = await patientRest.fetchPatientsInfo();
    patientMap.clear();
    for (var patient in restList) {
      putPatient(patient);
    }
    _logger.logDebug(1, "Total patients fetched: ${restList.length}");
    return patientMap.values.toList();
  }

  Future<List<Patient>> updatePatients(List<Patient> changedPatients) async {
    _logger.logTrace(2, 'Updating patients in the REST API');
    try {
      List<Patient> updatedPatients =
          await patientRest.putUpdatePatients(changedPatients);
      for (var patient in updatedPatients) {
        putPatient(patient); // Update local cache with new patient data
      }
      _logger.logTrace(2, "Successfully updated patients");
      return updatedPatients;
    } catch (e) {
      _logger.logError(2, 'Error updating patients: $e');
      throw Exception("Error attempting to update patients");
    }
  }

  Future<void> deletePatients(List<Patient> patientsToDelete) async {
    List<String> patientIds =
        patientsToDelete.map((patient) => patient.id).toList();
    _logger.logTrace(2, 'Deleting patients from REST API: $patientIds');

    for (String id in patientIds) {
      await patientRest.deletePatient(id);
      patientMap.remove(id);
    }

    _logger.logDebug(
        2, "${patientsToDelete.length} patients deleted successfully.");
  }

  Future<List<Patient>> refreshFromServer() async {
    _logger.logTrace(1, 'Refreshing patients from REST API (server truth)');
    final List<Patient> restList = await patientRest.fetchPatientsInfo();

    // Replace local cache with EXACT server set
    patientMap.clear();
    for (final p in restList) {
      putPatient(p);
    }
    _logger.logDebug(1, "Server returned ${restList.length} patients");
    return patientMap.values.toList();
  }
}

File: mpif/src/main/java/com/ivir/mpif/patient/db/PatientRepository.java
/*
 * Copyright 2026 IVIR Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ivir.mpif.patient.db;

import com.ivir.mpif.patient.PatientSource;
import jakarta.annotation.PostConstruct;
import org.dizitart.no2.Nitrite;
import org.dizitart.no2.collection.Document;
import org.dizitart.no2.collection.events.CollectionEventInfo;
import org.dizitart.no2.collection.events.CollectionEventListener;
import org.dizitart.no2.repository.ObjectRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.function.Consumer;

@Component
public class PatientRepository {
    private @Autowired Nitrite nitrite;
    private ObjectRepository<PatientEntity> repository;

    @PostConstruct
    public void init(){
        repository = nitrite.getRepository(PatientEntity.class);
    }

    public PatientEntity getById(String id) {
        return repository.getById(id);
    }

    public void update(PatientEntity patientEntity) {
        repository.update(patientEntity);
    }

    public List<PatientEntity> getAll(){
        return repository.find().toList();
    }

    public void insert(PatientEntity patientEntity){
        repository.insert(patientEntity);
    }

    public void remove(PatientEntity patientEntity){
        repository.remove(patientEntity);
    }

    public void addListener(Consumer<PatientEntity> newPatientConsumer,
                            Consumer<PatientEntity> updatedConsumer,
                            Consumer<PatientEntity> removedConsumer){
        repository.subscribe(new PatientEventListener(newPatientConsumer, updatedConsumer, removedConsumer));
    }

    private class PatientEventListener implements CollectionEventListener{
        private Consumer<PatientEntity> _newConsumer;
        private Consumer<PatientEntity> _updatedConsumer;
        private Consumer<PatientEntity> _removedConsumer;

        public PatientEventListener(Consumer<PatientEntity> newPatientConsumer,
                                    Consumer<PatientEntity> updatedConsumer,
                                    Consumer<PatientEntity> removedConsumer){
            _newConsumer = newPatientConsumer;
            _updatedConsumer = updatedConsumer;
            _removedConsumer = removedConsumer;
        }

        @Override
        public void onEvent(CollectionEventInfo<?> collectionEventInfo) {
            try {
                Document document = (Document) collectionEventInfo.getItem();
                PatientEntity patientEntity = new PatientEntity();
                patientEntity.setPatientCaseNum(document.get("patientCaseNum", Integer.class));
                patientEntity.setId(document.get("id", String.class));
                patientEntity.setPhysiologySource(toPatientSource(document.get("physiologySource")));
                patientEntity.setVisible(document.get("visible", Boolean.class));
                patientEntity.setSourceLocked(document.get("sourceLocked", Boolean.class));

                switch (collectionEventInfo.getEventType()) {
                    case Insert -> _newConsumer.accept(patientEntity);
                    case Remove -> _removedConsumer.accept(patientEntity);
                    default -> _updatedConsumer.accept(patientEntity);
                }
            }catch (Exception e){
                e.printStackTrace();
            }
        }
    }

    private PatientSource toPatientSource(Object value){
        if(value == null){
            return null;
        }
        return PatientSource.valueOf(value.toString());
    }
}
